Doc: Contribution Guide Updates (#3766)
aristanetworks/ansible-avd: ansible_collections/arista/avd/docs/contribution/development-tooling.md

<!--
  ~ Copyright (c) 2023-2024 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Development Tooling

- To assist the AVD development community, we provide guidance to develop with two primary methods: VSCode dev containers or local Python environment.

>   - You may choose your also choose your own development methodology, however we may not be able to provide assistance in a timely manner.

Reviewer: Repeated statements here...
"You may choose your also choose your own development methodology"

- This guide provides additional information about the development tools leverage in the project: pre-commit, Molecule, ansible-test
- Please report any issues and optimization suggestions regarding the development workflow via [Github discussions board](https://github.com/aristanetworks/avd/discussions).

## Development environments

### VSCode Dev Containers

- To facilitate onboarding of development the AVD project builds [Dev Containers](../containers/overview.md) with all the required tools to get started.
- Before you can leverage the Dev Container ensure to have [Docker](https://docs.docker.com/engine/install/) and [Visual Studio Code](https://code.visualstudio.com/) installed.

> - Follow the instruction to customize a and leverage the [AVD Dev Container](../containers/overview.md#how-to-use-dev-containers)

Reviewer: "to customize a and leverage" seems wrong

Reviewer 2: Updated!

    make converge MOLECULE=eos_designs_unit_tests ANSIBLE_OPTIONS="-vvv --forks 10"

    # Run eos_cli_config_gen scenario with limit.
    make converge ANSIBLE_OPTIONS="--limit logging"
    ```

4. Review generated artifacts and test results on an EOS device to confirm syntax and working configuration.
5. Commit generated artifacts.

!!! info
    Molecule scenarios are also executed in the project CI to ensure code quality and compliance and tested against various versions of ansible-core and other Python dependencies.

### Advanced CLI syntax

You may also run Molecule by leveraging its CLI syntax directly from the root of the collection path: `ansible_collections/arista/avd`.

Examples:

```shell
# Change to root collection directory
cd ansible_collections/arista/avd

# Run eos_cli_config_gen scenario

molecule converge -s eos_cli_config_gen

# Run eos_cli_config_gen scenario limiting to "logging" host only

molecule converge -s eos_cli_config_gen -- --limit logging

# Run eos_designs)unit_test scenario with verbosity
molecule converge -s eos_designs_unit_tests -- -vvv
```

## Ansible-test

> - The AVD project leverages [ansible-test](https://www.ansible.com/blog/introduction-to-ansible-test/) to test the asrista.avd Ansible Collection for sanity, unit and integration tests.

Reviewer: arista.avd :)

- Additionally, the AVD project leverages pre-commit hooks to build and update the AVD schemas and documentation artifacts.

### Install pre-commit hook

Configure pre-commit git hook, to automatically run pre-commit. This is optional, but highly recommended!

```shell
# Change to directory to your cloned avd repository.
cd avd

# Install pre-commit hooks to run automatically.
pre-commit install
```

### Run pre-commit manually

To run `pre-commit` manually before you commit, use this command:

```shell
# Run pre-commit hooks on all staged files.
# The command will automatically detect changed files using `git status` and run tests according to their type.
pre-commit run

# Run pre-commit hooks on all un-staged and staged files.
pre-commit run --all

# Run specific pre-commit schemas hook on all un-staged and staged files.
pre-commit run schemas --all
```

!!! note
    This process is also implemented in the project CI to ensure code quality and compliance.
    All pre-commit checks must pass, therefore we highly recommend running this workflow before committing changes!

## Molecule

> - [Molecule](https://ansible.readthedocs.io/projects/molecule/) project is designed to aid developing and testing of Ansible roles.
> - The AVD project leverages Molecule for:
>   - Static integration test on the following Ansible roles:
>     - `eos_designs`
>     - `eos_cli_config_gen`
>     - `eos_validate_state`
>     - `eos_config_deploy_cvp`
>     - `dhcp_provisioner`
>   - End-to-end systems integration tests on the following CloudVision role and module:
>     - `cv_deploy`
>     - `cv_workflow`
> - The molecule scenarios are located under the [molecule directory](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) at the root of the collection.
>   - The directory name is the molecule `--scenario-name` leveraged when executing Molecule, i.e: `eos_cli_config_gen`, `eos_designs_unit_tests`.

Reviewer 3: ```suggestion
The [Molecule](https://ansible.readthedocs.io/projects/molecule/) project is designed to aid in developing and testing of Ansible roles.

The AVD project leverages Molecule for:

- Static integration test on the following Ansible roles:
  - `eos_designs`
  - `eos_cli_config_gen`
  - `eos_validate_state`
  - `eos_config_deploy_cvp`
  - `dhcp_provisioner`
- End-to-end systems integration tests on the following CloudVision role and module:
  - `cv_deploy`
  - `cv_workflow`

The Molecule scenarios are located under the [`molecule` directory](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) at the root of the collection.

The directory name of each Molecule scenario folder is used as the `--scenario-name` when executing Molecule, i.e: `eos_cli_config_gen`, `eos_designs_unit_tests`.
```

  - This is the recommended way for development, as it is quicker and does not execute idempotency checks.
- `test`: Execute molecule "test" sequence. Specify scenario name with `MOLECULE=<scenario_name>` (default: `eos_cli_config_gen`) and Ansible options with `ANSIBLE_OPTIONS=<options>` (default: `--forks 5`).
  - This is executed as part of the CI and tests for idempotency.
- `refresh-facts`: Run all "eos_designs" and "eos_cli_config_gen" [molecule scenarios](https://github.com/aristanetworks/avd/blob/devel/ansible_collections/arista/avd/molecule/MOLECULE_SCENARIOS.txt).

!!! info
    `make refresh-facts` can be useful when your change common template or structured configuration output.
    Note that it will take a significant amount of local resources and several minutes/hours to execute.

In the majority of new features or bug fixes, the process is the following:

1. Update scenario inventory when required. It is ok to extend an existing host to cover a new test. When in doubt consult with a maintainer.
2. Update group and or host variables in the scenario and ensure to cover all use cases of the feature.

> 3. Run `make converge` target within the molecule directory to generate artifacts. Examples:

Reviewer 3: ```suggestion
3. Run `make converge` target within the `molecule` directory to generate artifacts. Examples:
```

    ```shell
    # Change to molecule directory
    cd ansible_collections/arista/avd/molecule

    # Run eos_designs_unit_tests scenario to generate artifacts
    make converge MOLECULE=eos_designs_unit_tests

    # Run eos_designs_unit_tests scenario with verbosity `-vvv` and max forks of 10 `--forks 10`.
    make converge MOLECULE=eos_designs_unit_tests ANSIBLE_OPTIONS="-vvv --forks 10"

    # Run eos_cli_config_gen scenario with limit.
    make converge ANSIBLE_OPTIONS="--limit logging"
    ```

4. Review generated artifacts and test results on an EOS device to confirm syntax and working configuration.
5. Commit generated artifacts.

!!! info

>     Molecule scenarios are also executed in the project CI to ensure code quality and compliance and tested against various versions of ansible-core and other Python dependencies.

Reviewer 3: ```suggestion
    Molecule scenarios are also executed in the project CI to ensure code quality and compliance and tested against various versions of `ansible-core` and other Python dependencies.
```

### Executing Molecule with advanced CLI syntax

You may also run Molecule by leveraging its CLI syntax directly from the root of the collection path: `ansible_collections/arista/avd`.

Examples:

```shell
# Change to root collection directory
cd ansible_collections/arista/avd

# Run eos_cli_config_gen scenario
molecule converge -s eos_cli_config_gen

# Run eos_cli_config_gen scenario limiting to "logging" host only
molecule converge -s eos_cli_config_gen -- --limit logging

# Run eos_designs)unit_test scenario with verbosity
molecule converge -s eos_designs_unit_tests -- -vvv
```

## Ansible-test

> - The AVD project leverages [ansible-test](https://www.ansible.com/blog/introduction-to-ansible-test/) to test the asrista.avd Ansible Collection for sanity, unit and integration tests.
> - Testing is performed automatically as part of the CI pipeline.
> - If troubleshooting is required, a Makefile at the root of the avd repository supports the following targets to execute ansible-test:
>   - `sanity`: Run ansible-test sanity validation.
>   - `unit-tests`: Run unit test cases using ansible-test. Specify `ANSIBLE_TEST_MODE=<docker|venv>` (default: `docker`).
>   - `integration-tests`: Run integration test cases using ansible-test. Specify `ANSIBLE_TEST_MODE=<docker|venv>` (default: `docker`).

Reviewer 3: ```suggestion
The AVD project leverages [ansible-test](https://www.ansible.com/blog/introduction-to-ansible-test/) to run sanity, unit and integration tests for the `arista.avd` Ansible Collection.

Testing is performed automatically as part of the CI pipeline. If troubleshooting is required, a Makefile at the root of the `avd` repository supports the following targets to execute `ansible-test`:

- `sanity`: Run ansible-test sanity validation.
- `unit-tests`: Run unit test cases using ansible-test. Optionally specify `ANSIBLE_TEST_MODE=<docker|venv>` (default: `docker`).
- `integration-tests`: Run integration test cases using `ansible-test`. Optionally specify `ANSIBLE_TEST_MODE=<docker|venv>` (default: `docker`).
```

<!--
  ~ Copyright (c) 2023-2024 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Development Tooling

- To assist the AVD development community, we provide guidance to develop with two primary methods: VSCode dev containers or local Python environment.
  - You may also choose your own development methodology, however we may not be able to provide assistance in a timely manner.
- This guide provides additional information about the development tools leveraged in the project: pre-commit, Molecule and ansible-test.
- Please report any issues and optimization suggestions regarding the development workflow via [Github discussions board](https://github.com/aristanetworks/avd/discussions).

!!! note

>     This guide doesn't got into details on how to install common development tooling as this differs from operating system to operating system.

Reviewer 4: syntax typo?
```suggestion
    This guide doesn't go into details on how to install common development tooling as this differs from operating system to operating system.
```

- [Python 3.10](https://docs.python.org/) or later and [Python Virtual Environments](https://docs.python.org/3/tutorial/venv.html).
- Additional AVD Python package dependencies.
- [Make](https://www.gnu.org/software/make/manual/make.html): Leveraged for automating software building and test procedures.
- [Docker](https://docs.docker.com/engine/install/) (Optional): Some of the tests require docker to be executed locally and useful when troubleshooting failures of the CI pipeline.

Recommended steps with Python virtual environment:

1. Create and activate a Python virtual environment.
2. Install Python requirements located in the AVD repository: [requirements-dev.txt](https://github.com/aristanetworks/avd/blob/devel/ansible_collections/arista/avd/requirements-dev.txt) and [requirements.txt](https://github.com/aristanetworks/avd/blob/devel/ansible_collections/arista/avd/requirements.txt).

!!! note
    Ensure the virtual environment is located outside of the AVD project directory.

```shell
# Create a Python virtual environment `python -m venv <virtual-environment-name>`.

> python3 -m venv avd-venv

Reviewer 4: Should we explicitly mention somewhere that all examples are valid for Linux/MacOS but not for Windows (as path syntax is different, venv activated differently, etc)?

Author: good point, I'll make that statement at the beginning.

Author: addressed in [8f26bf5](https://github.com/aristanetworks/avd/pull/3766/commits/8f26bf567b2ab09842ddade102dbff6c64ee6539)

# Activate Python virtual environment `source <virtual-environment-name>/bin/activate`.
source avd-venv/bin/activate

# Install AVD project requirements-dev.txt and requirements.txt in your Python Virtual environment.
# Requirements files are located in `ansible_collections/arista/avd` of the avd repository.

> pip3 install -r ansible_collections/arista/avd/requirements-dev.txt -r ansible_collections/arista/avd/requirements.txt --upgrade

Reviewer 4: Steps on lines 66 and 69 assume we are in the 'root' folder with subfolders 'avd' and 'avd-venv'

```suggestion
pip3 install -r avd/ansible_collections/arista/avd/requirements-dev.txt -r avd/ansible_collections/arista/avd/requirements.txt --upgrade
```

### Executing Molecule with makefile method

To run the Molecule tests locally to generate the new expected configuration and documentation leverage Makefile located in the `ansible_collections/arista/avd/molecule` directory.

The Makefile supports the following targets:

- `help`: Display available make target and descriptions.
- `converge`: Execute molecule "converge" sequence. Specify scenario name with `MOLECULE=<scenario_name>` (default: `eos_cli_config_gen`) and Ansible options with `ANSIBLE_OPTIONS=<options>` (default: `--forks 5`).
  - This is the recommended way for development, as it is quicker and does not execute idempotency checks.
- `test`: Execute molecule "test" sequence. Specify scenario name with `MOLECULE=<scenario_name>` (default: `eos_cli_config_gen`) and Ansible options with `ANSIBLE_OPTIONS=<options>` (default: `--forks 5`).
  - This is executed as part of the CI and tests for idempotency.
- `refresh-facts`: Run all "eos_designs" and "eos_cli_config_gen" [molecule scenarios](https://github.com/aristanetworks/avd/blob/devel/ansible_collections/arista/avd/molecule/MOLECULE_SCENARIOS.txt).

!!! info

>     `make refresh-facts` can be useful when your change common template or structured configuration output.

Reviewer 4: ```suggestion
    `make refresh-facts` can be useful when you change common template or structured configuration output.
```

# Run eos_cli_config_gen scenario
molecule converge -s eos_cli_config_gen

# Run eos_cli_config_gen scenario limiting to "logging" host only
molecule converge -s eos_cli_config_gen -- --limit logging

# Run eos_designs unit_test scenario with verbosity
molecule converge -s eos_designs_unit_tests -- -vvv
```

## Ansible-test

The AVD project leverages [ansible-test](https://www.ansible.com/blog/introduction-to-ansible-test/) to run sanity, unit and integration tests for the `arista.avd` Ansible Collection.

Testing is performed automatically as part of the CI pipeline. If troubleshooting is required, a Makefile at the root of the `avd` repository supports the following targets to execute `ansible-test`:

- `sanity`: Run ansible-test sanity validation.
- `unit-tests`: Run unit test cases using ansible-test. Optionally specify `ANSIBLE_TEST_MODE=<venv|docker>` (default: `venv`).
- `integration-tests`: Run integration test cases using `ansible-test`. Optionally specify `ANSIBLE_TEST_MODE=<venv|docker>` (default: `venv`).

Examples:

```shell
# Run ansible-test sanity validation. Requires docker.
make sanity

# Run unit test cases using ansible-test with venv (default)

> make unit-test

Reviewer 4: ```suggestion
make unit-tests
```

<!--
  ~ Copyright (c) 2023-2024 Arista Networks, Inc.
  ~ Use of this source code is governed by the Apache License 2.0
  ~ that can be found in the LICENSE file.
  -->

# Development Tooling

## Makefile
<!-- TO DO - update!!! -->

## Pre-commit hook

[pre-commit](https://github.com/aristanetworks/avd/blob/devel/.pre-commit-config.yaml) can run standard hooks on every commit to automatically point out issues in code such as missing semicolons, trailing whitespace, and debug statements. Pointing these issues out before code review allows a code reviewer to focus on the architecture of a change while not wasting time with trivial style nitpicks.

> 

Reviewer: The CI will perform the same checks which must be passed, so it's better for you if you run those tests and fix issues before submitting any changes.

Author: see commit 9b2bc1055ae33674b4288c2d640628ed4cc0ee4d

Repository implements the following hooks:

- `trailing-whitespace`: Fix trailing whitespace. If found, the commit is stopped, and you must rerun the commit process.
- `end-of-file-fixer`: Like `trailing-whitespace`, this hook fixes the wrong end of the file and stops your commit.
- `check-yaml`: Checks that all YAML files are valid.
- `check-added-large-files`: Check if no large file is included in the repository.
- `check-merge-conflict`: Validate there is no `MERGE` syntax related to an invalid merge process.
- `pylint`: Run Python linting with settings defined in [pylintrc](https://github.com/aristanetworks/avd/blob/devel/pylintrc).
- `yamllint`: Validate all YAML files using configuration from [yamllintrc](https://github.com/aristanetworks/avd/blob/devel/.github/yamllintrc).
- `ansible-lint`: Validate YAML files with Ansible proven practices, patters, and behaviors.
- `Flake8`: Style guide enforcement for Python code base.
- `markdownlint-cli`: Validates markdown files for common errors as referenced [here](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md).

> 

Reviewer 5: j2lint also ?

Author: we have decided to simply point to the pre-commit-config.yaml file so we don't have to maintain a list!
